akka-remote/src/main/scala/akka/remote/artery/compress/HeavyHitters.scala

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (C) 2009-2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.remote.artery.compress
+
+import scala.collection.immutable
+
+// TODO describe / think about how we could handle an actor dying and removing from "top"
+// TODO this is the scenario when "super traffic actor" goes away and we want to not use up the slot in HH
+
+/**
+ * INTERNAL API
+ *
+ * Mutable.
+ *
+ * Keeps a number of specific heavy hitters around in memory.
+ *
+ * See also Section 5.2 of http://dimacs.rutgers.edu/~graham/pubs/papers/cm-full.pdf
+ * for a discussion about the assumptions made and guarantees about the Heavy Hitters made in this model.
+ * We assume the Cash Register model in which there are only additions, which simplifies HH detecion significantly.
+ */
+private[remote] sealed class HeavyHitters[T](strategy: HeavyHittersStrategy[T]) {
+  // TODO technically could be optimised depending on strategy (priority queue?) (store in order of counts, drop last if exceeded size)
+  private[this] var _items = List.empty[HeavyHitter[T]]
+
+  def items: immutable.Set[T] = _items.map(_.item)(scala.collection.breakOut).toSet // TODO avoid the toSet? does anyone care?
+
+  /**
+   * Attempt adding item to heavy hitters set, if it does not fit in the top yet,
+   * it will be dropped and the method will return `false`.
+   *
+   * @return `true` if the added item is indeed a heavy hitter
+   */
+  def update(item: T, count: Long): Boolean =
+    _items.find(_.item == item) match { // TODO optimise? could it be the same case always?
+      case Some(existing) ⇒
+        // if item already was a heavy hitter, and since we are in an addition only mode, no need to check counts,
+        // as the new count always will be at-least 1 more than previously (e.g.g new message arrived so we increment the count
+        // and end up checking if it is a HH).
+        _items.filterNot(_ == existing) // TODO maybe we could avoid iterating over here again (change List to indexable structure) // needs bench
+        _items ::= HeavyHitter(item, count)
+        false // already was registered
+      case None ⇒
+        // item wasn't yet a heavy hitter, thus we need to insert it and trim out any HH that do not fit the range anymore
+        val incoming = HeavyHitter(item, count)
+        _items = strategy.findAll(incoming :: _items)
+        _items contains incoming
+    }
+
+  override def toString =
+    s"${getClass.getSimpleName}(\n" +
+      s"  ${_items.mkString(",\n  ")}\n" +
+      s")"
+}
+
+/**
+ * INTERNAL API
+ * Holder for idenfified heavy hitter along with it's occurance count (often aproximated)
+ */
+private[remote] final case class HeavyHitter[T](item: T, count: Long)
+
+/**
+ * INTERNAL API
+ * [[HeavyHitters]] that never triggers a detection.
+ */
+private[remote] object NeverHeavyHitters extends HeavyHitters[Any](null) {
```

Comment:
There is no difference between topPercentage or topN apart from that topN keeps the table bounded while percentage does not. Since we want to keep the table advertisements bounded, topN is the natural choice I think.